Add easy global proxy support

msaong opened the pull request:
1. Let electron to support proxy via setting file, the file will auto create in the APPDATA, anyone can edit the file to add/change proxy setting.
2. After change the setting, user must restart the APP.
3. Add easy electron menu to show the proxy state.
4. Test and pass in macbook m1 chips

<img width="189" alt="image" src="https://github.com/sunner/ChatALL/assets/25732857/fc442cdc-ac4b-4b3e-b885-e1252b90785f">
<img width="797" alt="image" src="https://github.com/sunner/ChatALL/assets/25732857/d07ba500-957f-4e3a-97a0-d12e3acebbc6">
<img width="797" alt="image" src="https://github.com/sunner/ChatALL/assets/25732857/1411338d-bb88-40be-b54e-27dd63359973">

sunner commented:
Thank you @msaong . Proxy is an important feature for ChatALL. Appreciate you work.

I tested your code and meet some problems:

1. You mentioned the .json file will be auto created. But I can't find any code about creating and get the following error when launching:
```
Read proxy file failed: Error: ENOENT: no such file or directory, open '/Users/xxxxxxx/Library/Application Support/chatall/proxySetting.json'
App threw an error during load
TypeError: Cannot read properties of undefined (reading 'proxyServer')
    at eval (webpack://chatall/./src/background.js?:35:18)
```

2. From you code, I learned that the command line params can change proxy. After reading https://www.electronjs.org/docs/latest/api/command-line-switches , I found `--proxy-pac-url` maybe is a better choice than `--proxy-server` if it supports local file. A PAC file can route different site to different proxy (http or socks) or direct connection. What do you think?

If you love the PAC method too, we can make a template and tell users how to edit it as they like.

Thank you agian.

msaong commented:
@sunner fix the proxy file auto created now. Please check again.

PAC function is good, i need some time to finish and do some test, just show the new proxy setting menu as below:
<img width="211" alt="image" src="https://github.com/sunner/ChatALL/assets/25732857/d9632207-0546-408b-92be-faf39cd6a219">

sunner commented:
Thank you for keeping improving.

PAC mode also supports proxy all requests. So, do we need a proxy mode named `All`?

msaong commented:
pac support is finish.

![image](https://github.com/sunner/ChatALL/assets/25732857/81e9a1fb-40fd-4975-a36b-c301d5be1e8f)

msaong commented:
PAC file example for test, it works, but pac file need to improve:

`function FindProxyForURL(url, host) {
    if (shExpMatch(host, "*.aliyun.com") ||
        shExpMatch(host, "*.tiangong.cn") ||
        shExpMatch(host, "*.xfyun.cn") ||
        shExpMatch(host, "*.baidu.com") ||
        shExpMatch(host, "*.baidubce.com")) {
        return "DIRECT";
    }
    // No proxy for localhost
    if (isInNet(dnsResolve(host), "127.0.0.0", "255.0.0.0")) {
        return "DIRECT";
    }
    return "PROXY 127.0.0.1:7890; DIRECT";

}`

sunner commented:
> pac support is finish.
> 
> ![image](https://user-images.githubusercontent.com/25732857/248637545-81e9a1fb-40fd-4975-a36b-c301d5be1e8f.png)

Do you want to implement modifying PAC content in the menus? Perhaps it is hard and lost flexibility. User may use different flavor of PAC.

sunner commented:
> PAC file example for test, it works, but pac file need to improve [samplepac.txt](https://github.com/sunner/ChatALL/files/11862783/samplepac.txt)

As I know, the most popular template of PAC is adblockplus. Its document: https://help.adblockplus.org/hc/en-us/articles/360062733293

The attachment is used by me. 
[proxy.js.txt](https://github.com/sunner/ChatALL/files/11863002/proxy.js.txt)

msaong commented:
> > pac support is finish.
> > ![image](https://user-images.githubusercontent.com/25732857/248637545-81e9a1fb-40fd-4975-a36b-c301d5be1e8f.png)
> 
> Do you want to implement modifying PAC content in the menus? Perhaps it is hard and lost flexibility. User may use different flavor of PAC.

It's impossible to modifying PAC content in the electron menus, but add vue files for this purpose is ok. 

I use Electron Menu to implement simple proxy management just simply feel that the main code of the current program and this proxy function are independent of each other and easy to maintain

msaong commented:
@sunner I make some improve, it's much better

1. Added proxy configuration feature in the program settings menu"
2. Removed Electron proxy menu
3. Added automatic detection of proxy configuration file at program startup"

https://github.com/sunner/ChatALL/assets/25732857/13275873-7feb-463f-97ca-fdbfd62b6a1c

sunner commented on src/i18n/locales/en.json:
```diff
@@ -160,5 +160,34 @@
+  "proxy":{
+    "name": "Proxy",
+    "address": "Proxy Server Address",
+    "addressHint":"Proxy server address, e.g., http://127.0.0.1:7890",
```
```suggestion
    "addressHint":"HTTP proxy only, e.g., http://127.0.0.1:7890",
```

sunner commented on src/i18n/locales/en.json:
```diff
@@ -160,5 +160,34 @@
+    "name": "Proxy",
+    "address": "Proxy Server Address",
+    "addressHint":"Proxy server address, e.g., http://127.0.0.1:7890",
+    "enable":"Enable",
+    "disable":"Disable",
```
There is an `settings.enable`.

sunner commented on src/i18n/locales/zh.json:
```diff
@@ -160,5 +160,34 @@
     "wrapAround": "循环查找"
+  },
+  "proxy":{
+    "name": "代理服务器配置",
```
```suggestion
    "name": "代理服务器",
```

sunner commented on src/i18n/locales/zh.json:
```diff
@@ -160,5 +160,34 @@
+  },
+  "proxy":{
+    "name": "代理服务器配置",
+    "address": "代理服务器地址",
```
```suggestion
    "address": "HTTP 代理服务器地址",
```

sunner commented on src/i18n/locales/zh.json:
```diff
@@ -160,5 +160,34 @@
+  "proxy":{
+    "name": "代理服务器配置",
+    "address": "代理服务器地址",
+    "addressHint":"代理服务器地址，如 http://127.0.0.1:7890",
```
```suggestion
    "addressHint":"只能使用 HTTP 代理，如 http://127.0.0.1:7890",
```

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,227 @@
+        {{ $t('proxy.reset') }}
+      </v-btn>
+    </v-list-item>
+    <v-list-item>
+        <v-alert
+    type="info"
+    variant="outlined"
+    :text="$t('proxy.saveHint')"
+  ></v-alert>
+    </v-list-item>
```
```suggestion

```
Using `ConfirmModal.vue` to tell this info and ask for confirmation may by better.

But the `v-alert` is good for ChatGPT setting. I will use it there.

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,227 @@
+        {{ $t('proxy.onlySave') }}
+      </v-btn>
+      <v-btn
+        color="#5865f2"
```
```suggestion
        color="primary"
```

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,227 @@
+    <v-list-item>
+      <v-list-item-title>{{ $t('proxy.action') }}</v-list-item-title>
+      <v-btn
+        color="#5865f2"
```
```suggestion
        variant="outlined"
        color="primary"
```

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,227 @@
+        {{ $t('proxy.saveAndApply') }}
+      </v-btn>
+      <v-btn
+        color="#5865f2"
```
```suggestion
        variant="outlined"
        color="primary"
```

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,227 @@
+        {{ $t('proxy.reload') }}
+      </v-btn>
+      <v-btn
+        color="#5865f2"
```
```suggestion
        variant="outlined"
        color="primary"
```

sunner reviewed (changes requested):
Your PR is really great! More than what I can do. Thank you very much!

There are several style mismatch. I commented them inline.

IMO, the UI needs to be changed much. The following is my design.

<img width="811" alt="image" src="https://github.com/sunner/ChatALL/assets/255413/1584d6a7-ad54-480e-863d-ae9e35481a40">

Is it more clear for new users?

Eager for your next commit and merging them!

msaong commented:
@sunner Commit Again!

1. Optimized code and UI based on provided suggestions.
2. Implemented a convenient way to quickly set proxies based on models.
3. Added confirmation interactions for 'Restart Program', 'Reload Configuration', and 'Reset to Default' functions for better operation assurance.
4. Cleaned up the previous menu.js file.

You mentioned "The assertion that the proxy can only use the HTTP protocol“ is incorrect. Electron does not limit the use to HTTP proxy only; I can use a SOCKS proxy just as effectively. Therefore, I will retain the original description in the corresponding text.

![image](https://github.com/sunner/ChatALL/assets/25732857/ac08c9db-d2cf-43e9-858b-b2d9ecd7d0a7)

sunner commented:
> You mentioned "The assertion that the proxy can only use the HTTP protocol“ is incorrect. Electron does not limit the use to HTTP proxy only; I can use a SOCKS proxy just as effectively. Therefore, I will retain the original description in the corresponding text.

Great! I must misunderstand Electron docs. So how to setup a SOCKS proxy? socks://127.0.0.1:1080 ? If so, give an example is helpful.

msaong commented:
> > You mentioned "The assertion that the proxy can only use the HTTP protocol“ is incorrect. Electron does not limit the use to HTTP proxy only; I can use a SOCKS proxy just as effectively. Therefore, I will retain the original description in the corresponding text.
> 
> Great! I must misunderstand Electron docs. So how to setup a SOCKS proxy? socks://127.0.0.1:1080 ? If so, give an example is helpful.

The Hint already add the socks' sample
![image](https://github.com/sunner/ChatALL/assets/25732857/88532d4b-1f1f-4de7-9a53-5ebd88fe65f8)

sunner commented:
Thank you very much.

Could you resolve the conflicts? A rebasing with the main branch is better.

Now, the [files changed](https://github.com/sunner/ChatALL/pull/276/files) includes many diffs which are distraction.

msaong commented:
Conflicts resolve.

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,349 @@
+    name: i18n.global.t("wenxinQianfan.name"),
+    bypassList: "*.aip.baidubce.com",
+  },
+  {
+    name: i18n.global.t("dev.name"),
+    bypassList: "*.chatall.a",
+  },
```
```suggestion
```
The dev bot is for debugging only.

sunner commented on src/background.js:
```diff
@@ -1,14 +1,89 @@
+const proxySettingPath = path.join(userDataPath, 'proxySetting.json');
+const defaultProxySetting = {
+  enableProxy: false,
+  proxyMode: "All",
```
I suggest the three modes named: 

normal, Normal mode
pacFile, PAC file mode
pacUrl, PAC URL mode

"Global" may be miss understand as the system wide proxy setting. What do you think?

sunner commented on src/bots/openai/ChatGPT4Bot.js:
```diff
@@ -29,7 +29,6 @@ export default class ChatGPT4Bot extends ChatGPTBot {
     store.state.chatgpt.riskConfirmed = confirmed;
     return confirmed;
   }
-
```
Why?

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,349 @@
+    <v-list-item>
+      <v-list-item-title>{{ $t("proxy.enableProxy") }}</v-list-item-title>
+      <v-checkbox
+        :label="$t('proxy.enable')"
```
```suggestion
        :label="$t('settings.enable')"
```

sunner commented on src/components/ProxySetting.vue:
```diff
@@ -0,0 +1,349 @@
+        variant="outlined"
+        color="primary"
+        @click="resetAll"
+        class="ma-2 pa-2"
```
```suggestion
```

sunner commented on src/components/SettingsModal.vue:
```diff
@@ -6,15 +6,37 @@
         <v-toolbar-title>{{ $t("settings.title") }}</v-toolbar-title>
         <v-spacer></v-spacer>
-        <v-btn icon dark @click="closeDialog">
+        <v-tabs
```
Could you set the tabs to be vertical on the left?

sunner commented on src/i18n/locales/en.json:
```diff
@@ -160,5 +161,41 @@
+    "name": "Proxy",
+    "address": "Proxy Server Address",
+    "addressHint":"Proxy Server Address, e.g., http://127.0.0.1:7890 socks5://127.0.0.1:7890",
+    "enable":"Enable",
+    "disable":"Disable",
```
```suggestion
```

sunner commented on src/i18n/locales/zh.json:
```diff
@@ -160,5 +161,41 @@
+    "name": "代理服务器",
+    "address": "代理服务器地址",
+    "addressHint":"代理服务器地址，如 http://127.0.0.1:7890 socks5://127.0.0.1:7890",
+    "enable":"启用",
+    "disable":"禁用",
```
```suggestion
```

sunner reviewed (changes requested):
Every submit has surprise. Users must love the connivence provided by you.

msaong commented:
Optimized finish

![image](https://github.com/sunner/ChatALL/assets/25732857/efe413d2-0bc6-4cdc-b94a-e1b20f490dd4)

sunner commented:
Still can't rebase.......Can you resolve this? Or leave it to me.

msaong commented:
> Still can't rebase.......Can you resolve this? Or leave it to me.

I dont't know what happened. I compare two branch, it seems anything is ok. Any error message ?

sunner commented:
`git rebase origin/main` will get many conflicts.

msaong commented:
> `git rebase origin/main` will get many conflicts.

I reviewed the code and found that the conflicts stem from several files that were updated in the main branch today. These files are unrelated to my code. However, I'm not very familiar with Git commands, so resolving this issue would take me a significant amount of time. I think it would be faster if you handled the merge conflict problem.

sunner commented:
Thank you very much. The PR will be included in next release.

sunner commented:
@msaong I noticed that you are not listed in the contributors. Your code were commit with email ”1215927@qq.com“. Is it your GitHub register email or bind email. If not, bind it at https://github.com/settings/emails, then the commit will be counted on you.

msaong commented:
> @msaong I noticed that you are not listed in the contributors. Your code were commit with email ”[1215927@qq.com](mailto:1215927@qq.com)“. Is it your GitHub register email or bind email. If not, bind it at https://github.com/settings/emails, then the commit will be counted on you.

Add the QQ mail now.

sunner commented:
Looks like it takes time to take effect
